### PR TITLE
Add support for variables in value bindings

### PR DESCRIPTION
**Pull request by exyi**
The syntax is `val variableName = value; nextExpression`.
The variable is not assignable again - it is always immutable.
Variable may shadow property names and variables defined ahead of it.
The variable may be accessed only in the `nextExpression`.

This version does not support automatic unwrapping Tasks.
We wanted to switch to using await in DotVVM 3 anyway, right?

see #776

### CODE REVIEW

**Comment by exyi** [comment]
The parser/resolver and value binding work is done. The PR is marked as WIP because static commands are not working right now.

**Comment by exyi** [comment]
It now should work even in static commands. Arguably, the translation is a bit messy, but it works - we can probably improve the output JS code later.

**Review by acizmarik — APPROVED** [review]
Nice 👍

File: src/DotVVM.Framework.Tests.Common/Parser/Binding/BindingParserTests.cs [hunk 1]
```diff
@@ -934,6 +934,29 @@ public void BindingParser_MultiblockExpression_EmptyBlockMiddle_Invalid(string b
             Assert.AreEqual(node.EndPosition, lastExpression.EndPosition);
             Assert.AreEqual(voidBlockExpectedLenght, middleExpression.Length);
 
+        }
+        [DataRow("var x=A(); !x", "x", DisplayName = "Variable (val) expression")]
```
[the remark this answers is not in the document]
Comment (quigamdev) [reply src/DotVVM.Framework.Tests.Common/Parser/Binding/BindingParserTests.cs]: Fixed.

File: src/DotVVM.Framework/Compilation/Binding/StaticCommandBindingCompiler.cs [hunk 2]
```diff
@@ -50,6 +50,62 @@ public JsExpression CompileToJavascript(DataContextStack dataContext, Expression
...
+                        var tmpVar = new JsTemporaryVariableParameter();
+                        return Expression.Parameter(v.Type, v.Name).AddParameterAnnotation(new BindingParameterAnnotation(extensionParameter:
+                            new JavascriptTranslationVisitor.FakeExtensionParameter(_ => new JsSymbolicParameter(tmpVar))
+                        ));
+                    }).ToArray()
+                );
+
+                return core(replacedVariables);
+
+                // var fakeBlock = Expression.Block(
```
[the remark this answers is not in the document]
Comment (quigamdev) [reply src/DotVVM.Framework/Compilation/Binding/StaticCommandBindingCompiler.cs]: I have removed it. It seems like zombie code.

**Review by quigamdev — APPROVED** [review]

**Pull request merged**